presto-main/src/main/java/com/facebook/presto/operator/WindowOperator.java

```diff
@@ -924,17 +924,19 @@ static int findEndPosition(int startPosition, int endPosition, BiPredicate<Integ
         checkArgument(startPosition < endPosition, "startPosition (%s) must be less than endPosition (%s)", startPosition, endPosition);
 
         int left = startPosition;
-        int right = endPosition;
-
-        while (left + 1 < right) {
-            int middle = (left + right) >>> 1;
-
-            if (comparator.test(startPosition, middle)) {
-                left = middle;
-            }
-            else {
-                right = middle;
+        int right = left + 1;
+
+        while (right < endPosition) {
+            int distance = 1;
+            for (; distance < endPosition - left; distance *= 2) {
+                right = left + distance;
+                if (!comparator.test(left, right)) {
```

Comment:
I'm not super familiar with this code. But the previous code always does a comparison of a fixed value at startPosition with middle, whereas here both left and right are changing bounds and there isn't always a comparison with a fixed value so its hard to reason how this is a binary search. 

Please can you give some more explanation.

Reply:
Thanks for your comment, the code here is a bit subtle and I would give some more detailed explanation here:
- Although left and right are both changing bounds, the value of left bound always equals to the value at startPosition, because the if-condition would guarantee that the value[left] == value[left + distance / 2], whose base case is value[startPosition] == value[startPosition + distance / 2]. 
- The if-condition could be changed to 
```
if (!comparator.test(startPosition, right)) {
```
to ensure the comparison with the original value.
Both implementation would be correct, and I agree with your observation that changing both left bound and right bound not only seems subtle but also incurs more data reading.
So I would optimize it with the static left bound.